File: app/src/main/java/com/platdmit/simplecloudmanager/app/fragments/ServerTabBackupsFragment.java
package com.platdmit.simplecloudmanager.app.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.platdmit.simplecloudmanager.R;
import com.platdmit.simplecloudmanager.app.adapters.BackupListAdapter;
import com.platdmit.simplecloudmanager.app.vm.BackupsViewModel;
import com.platdmit.simplecloudmanager.app.vm.factory.SingleElementViewModelFactory;
import com.platdmit.simplecloudmanager.data.api.implement.ApiServerRepoImp;
import com.platdmit.simplecloudmanager.domain.SCMApp;
import com.platdmit.simplecloudmanager.domain.converters.implement.BackupConvertImp;
import com.platdmit.simplecloudmanager.domain.helpers.ContentUpdateService;
import com.platdmit.simplecloudmanager.domain.models.Backup;
import com.platdmit.simplecloudmanager.domain.repo.ServerBackupRepo;
import com.platdmit.simplecloudmanager.domain.repo.implement.ServerRepoImp;
import com.platdmit.simplecloudmanager.domain.repo.implement.UpdateScheduleRepImp;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class ServerTabBackupsFragment extends Fragment implements ServerTabFragment<ServerTabBackupsFragment>{
    private static final String TAG = ServerTabBackupsFragment.class.getSimpleName();
    private String mTitle;

    private BackupsViewModel mBackupsViewModel;
    private RecyclerView mRecyclerView;
    private BackupListAdapter mAdapter;

    public ServerTabBackupsFragment(){};

    public ServerTabBackupsFragment(String title) {
        mTitle = title;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        if(savedInstanceState != null){
            mBackupsViewModel = new ViewModelProvider(this).get(BackupsViewModel.class);
        } else {
            mBackupsViewModel = new ViewModelProvider(this,
                    new SingleElementViewModelFactory<ServerBackupRepo>(
                            new ServerRepoImp(
                                    new ApiServerRepoImp(SCMApp.getActualApiKeyService()),
                                    SCMApp.getDb(),
                                    new BackupConvertImp(),
                                    new ContentUpdateService(new UpdateScheduleRepImp(SCMApp.getDb()))
                            ), ServerBackupRepo.class, getArguments().getLong("ELEMENT_ID")
                    )).get(BackupsViewModel.class);
        }
        
        View v = inflater.inflate(R.layout.fragment_server_tab_backups, container, false);

        mRecyclerView = v.findViewById(R.id.server_backup_list);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mAdapter = new BackupListAdapter();

        mBackupsViewModel.getBackupsLiveData().observe(getViewLifecycleOwner(), this::updateServerActionData);

        return v;
    }

    @Override
    public String getTitle() {
        return mTitle;
    }

    @Override
    public ServerTabBackupsFragment getInstance() {
        return new ServerTabBackupsFragment();
    }

    private void updateServerActionData(List<Backup> backups){
        mAdapter.setContentData(backups);
        if(mRecyclerView.getAdapter() == null){
            mRecyclerView.setAdapter(mAdapter);
        } else {
            mAdapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/fragments/ServerTabBackupsFragment.kt
package com.platdmit.simplecloudmanager.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.platdmit.simplecloudmanager.R
import com.platdmit.simplecloudmanager.adapters.BackupListAdapter
import com.platdmit.simplecloudmanager.vm.BackupsViewModel
import com.platdmit.domain.models.Backup
import com.platdmit.simplecloudmanager.states.ActionState
import com.platdmit.simplecloudmanager.states.BackupsState
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.android.synthetic.main.fragment_server_tab_backups.*

@AndroidEntryPoint
class ServerTabBackupsFragment(
        private val mTitle: String = "empty"
) : Fragment(R.layout.fragment_server_tab_backups), ServerTabFragment<ServerTabBackupsFragment> {
    private val backupsViewModel: BackupsViewModel by viewModels()
    private val backupListAdapter = BackupListAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if(savedInstanceState == null){
            setStateIntent(BackupsViewModel.StateIntent.SetServerId(requireArguments().getLong("ELEMENT_ID")))
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        server_backup_list.layoutManager = LinearLayoutManager(context)

        backupsViewModel.backupsStateLiveData.observe(viewLifecycleOwner, Observer { stateHandler(it) })
    }

    override fun getTitle(): String {
        return mTitle
    }

    override fun getInstance(): ServerTabBackupsFragment {
        return ServerTabBackupsFragment()
    }

    private fun stateHandler(backupsState: BackupsState){
        when(backupsState){
            is BackupsState.Loading -> {}
            is BackupsState.Success -> {
                updateServerActionData(backupsState.backups)
            }
            is BackupsState.Empty -> {}
            is BackupsState.Error -> {}
        }
    }

    private fun setStateIntent(stateIntent: BackupsViewModel.StateIntent){
        backupsViewModel.setStateIntent(stateIntent)
    }

    private fun updateServerActionData(backups: List<Backup>) {
        backupListAdapter.setContentData(backups)
        if (server_backup_list.adapter == null) {
            server_backup_list.adapter = backupListAdapter
        } else {
            backupListAdapter.notifyDataSetChanged()
        }
    }
}